core/block_svg.js

```diff
@@ -619,9 +619,22 @@ Blockly.BlockSvg.prototype.tab = function(start, forward) {
   var target = list[forward ? i + 1 : i - 1];
   if (!target) {
     // Ran off of list.
-    var parent = this.getParent();
-    if (parent) {
-      parent.tab(this, forward);
+    // If there is an output, tab up to that block.
+    var outputBlock = this.outputConnection && this.outputConnection.targetBlock();
+    if (outputBlock) {
+      outputBlock.tab(this, forward);
+    } else { // Otherwise, go to next / previous block, depending on value of `forward`
+      if (forward) {
+        var nextBlock = this.nextConnection && this.nextConnection.targetBlock();
+        if (nextBlock) {
+          nextBlock.tab(this, forward);
+        }
+      } else {
+        var previousBlock = this.previousConnection && this.previousConnection.targetBlock();
+        if (previousBlock) {
+          previousBlock.tab(this, forward);
+        }
+      }
```

Comment:
Yeah that's what I figured.

I'm totally down with you merging this as-is, because it's an improvement over the existing code for any short-term uses of tabbing.  But it doesn't fix the problem all the way.  See what Erik said below.

There's probably a clever way to do it without using any external state, which would be preferable, but fixing it fully isn't a high priority for me right now.